### PR TITLE
Change from_bytes methods to take fixed-size array argument

### PR DESCRIPTION
**Pull request by mina86**
Change from_bytes methods of all the types to take `&[u8; N]` argument
(with `N` appropriate for given type) rather than `&[u8]`.  This does
a few things:
- it makes SecretKey::from_bytes infallible which will simplify call
  sites (if the call site already has properly-sized array),
- harmonises the convention with ed25519::Signature::from_bytes method
  and
- encodes the size of the bytes representation in type system which
  means that users can assert it at compile time.

To still allow creating the objects from a slice whose size is checked
at run time, introduce `TryFrom<&[u8]>` implementations for affected
public types.

This is an API breaking change.  The simplest way to update existing
code is to replace Foo::from_bytes with Foo::try_from.  This should
cover majority of uses.

### CODE REVIEW

File: src/keypair.rs [hunk 1]
```diff
@@ -49,6 +49,24 @@ impl From<SecretKey> for Keypair {
...
+        let secret = SecretKey::from(secret);
+        let public = PublicKey::try_from(public)?;
+
+        if public != (&secret).into() {
+            Err(InternalError::MismatchedKeypairError.into())
+        } else {
+            Ok(Keypair { secret, public })
+        }
+    }
+}
```
Comment (tarcieri) [inline src/keypair.rs:68]: This is copy/paste of `Keypair::from_bytes`:

https://github.com/dalek-cryptography/ed25519-dalek/blob/9638ab4/src/keypair.rs#L94-L110

At the very least one should be composed in terms of the other.

But also as I [discussed in #225](https://github.com/dalek-cryptography/ed25519-dalek/issues/225#issuecomment-1322358913), we should probably get rid of the 64-byte keypair APIs, or at least deprecate them, as they don't follow the RFC8032 definition of a secret key and include redundant information which provides no additional value and only permits potential misuse.

Comment (mina86) [reply src/keypair.rs:68]: > At the very least one should be composed in terms of the other.

They are.  This commit changes from_bytes to check length and then call try_from.

I actually don’t care about Keypair, but I figured that if I’m adding the conversion to SecretKey and PublicKey it makes sense for completeness to have it in Keypair as well.

**Comment by tarcieri** [comment]
In general I'd suggest holding off on this until #225 is addressed, since that proposes to change the public APIs for keys

**Comment by tarcieri** [comment]
This was also an interesting read on the actual costs of length checks: https://blog.readyset.io/bounds-checks/

tl;dr: not much, especially in cases like this

**Comment by mina86** [comment]
The main thing is that creation of SecretKey is now infallible.  Another is that the expected length is now expressed in type system.  Performance isn’t really my main concern.

**Comment by tarcieri** [comment]
I'd probably suggest making the `TryFrom` impls take `&[u8]` and using inherent methods (i.e. `from_bytes`) for the fixed-sized inputs. That's the approach we use with @RustCrypto.

**Comment by mina86** [comment]
I can change `from_bytes` if you prefer though that is an API breaking change.

**Comment by mina86** [comment]
So I’m looking at [ed25519::Signature::from_bytes from a preview ed25519 crate](https://docs.rs/ed25519/2.0.0-pre.0/ed25519/struct.Signature.html#method.from_bytes) and it has `from_bytes(&[u8])` so should this follow that instead? Or should ed25519 crate be changed as well?

**Comment by tarcieri** [comment]
I'd be open to changing that method

**Comment by mina86** [comment]
Updated as per suggestion.  Happy to wait till #225 gets resolved.

**Comment by tarcieri** [comment]
#225 is resolved

**Comment by tarcieri** [comment]
@mina86 you need to change the base to `release/2.0`

**Review by tarcieri — APPROVED** [review]
